### PR TITLE
Separation 3

### PR DESCRIPTION
**Pull request by micuami**

### CODE REVIEW

**Review by george-cosma — CHANGES_REQUESTED** [review]

File: tock-process-console/src/ui_management/pages/terminal_page/components/terminal_box.rs [hunk 1]
```diff
@@ -13,7 +13,8 @@ use crossterm::event::{KeyCode, KeyEventKind, MouseEventKind};
 use ratatui::{
     layout::{Alignment, Rect},
     style::{Color, Modifier, Style},
-    widgets::{Block, Borders, Paragraph},
+    widgets::block::Block,
+    widgets::{Borders, Paragraph},
```
Comment (george-cosma) [inline tock-process-console/src/ui_management/pages/terminal_page/components/terminal_box.rs:17]: What does this modification serve?

File: tockloader-lib/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+tokio-util = { version = "0.7.8", features = ["full"] }
+tock-process-console = { path = "../tock-process-console/" }
+probe-rs = "0.24.0"
+tbf-parser = { path = "../tbf-parser"}
+glob = "0.3.1"
+inquire = "0.7.5"
+utf8-decode = "1.0.1"
+byteorder = "1.5.0"
+tar = "0.4.41"
+
```
Comment (george-cosma) [inline tockloader-lib/Cargo.toml:28]: Are all of these dependencies needed?

File: tockloader-lib/src/attributes/app_attributes.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    pub app_version: Option<u32>,
+    pub tvl_package_type: Option<u16>,
+    pub package_name: Option<String>,
+    pub tvl_kernel_version_type: Option<u16>,
+    pub kernel_major: Option<u16>,
+    pub kernel_minor: Option<u16>,
+    pub kernel_version: Option<(u16, u16)>,
+    pub footer_size: Option<u32>,
+    pub address: Option<u64>,
+}
```
Comment (george-cosma) [inline tockloader-lib/src/attributes/app_attributes.rs:28]: Isn't this a TBF Object? Couldn't we use `tbf-parser` for this?

File: tockloader-lib/src/attributes/get_app_attributes.rs
Comment (george-cosma) [inline tockloader-lib/src/attributes/get_app_attributes.rs:1]: Why is this file seperate from `app_attributes`?

File: tockloader-lib/src/attributes/get_app_attributes.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        let (_ver, _header_len, whole_len) =
+            match parse_tbf_header_lengths(&buff[0..8].try_into().unwrap()) {
+                Ok((ver, header_len, whole_len)) if header_len != 0 => (ver, header_len, whole_len),
+                _ => break, // No more apps
+            };
+
+        address += whole_len as u64;
+    }
+    Some(address)
+}
```
Comment (george-cosma) [inline tockloader-lib/src/attributes/get_app_attributes.rs:146]: Is this function appropriate to put here? This doesn't seem relevant to `app_attributes`. Also, see comment attached to file

File: tockloader-lib/src/attributes/get_board_attributes.rs
Comment (george-cosma) [inline tockloader-lib/src/attributes/get_board_attributes.rs:1]: Why does this file exist? I would suggest merging it with `hardware_attributes.rs`

File: tockloader-lib/src/attributes/get_kernel_attributes.rs
Comment (george-cosma) [inline tockloader-lib/src/attributes/get_kernel_attributes.rs:1]: Why does this file exist? I would suggest merging it with `hardware_attributes.rs`

File: tockloader-lib/src/attributes/get_board_attributes.rs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+pub(crate) fn get_board_attributes(board_core: &mut Core, attributes: &mut HardwareAttributes) {
+    let address = 0x600;
+    let mut buf = [0u8; 64 * 16];
+    let _ = board_core.read(address, &mut buf);
+
+    let mut data = buf.chunks(64);
+
+    for index_data in 0..data.len() {
```
Comment (george-cosma) [inline tockloader-lib/src/attributes/get_board_attributes.rs:35]: What is index_data?

File: tockloader-lib/src/attributes/hardware_attributes.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    pub boothash: Option<String>,
+    pub bootloader_version: Option<String>,
+    pub sentinel: Option<String>,
+    pub kernel_version: Option<u64>,
+    pub app_mem_start: Option<u32>,
+    pub app_mem_len: Option<u32>,
+    pub kernel_bin_start: Option<u32>,
+    pub kernel_bin_len: Option<u32>,
+}
+
```
Comment (george-cosma) [inline tockloader-lib/src/attributes/hardware_attributes.rs:19]: I noticed that `HardwareAttributes` is actually a few different things:
- Kernel attributes (& Kernel version) from `get_kernel_attributes.rs`
- Board Attributes (& some other stuff) from `get_board_attributes.rs`

I would rather have two different structures. Eventually, these two structures alongside the vector of AppAttributes (which, again, could be replaced with `tbf-parser`) be bundled up in a generic `Attributes` struct

```rs
struct Attributes<'a> {
    bootloader: BootloaderAttributes;
    board: BoardAttributes;
    apps: Vec<TbfHeader<'a>>
}
```

File: tockloader-cli/src/display.rs [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+                    " \x1b[1;32m        Length:                 {:<10}\n\n",
+                    temp_data.footer_size.unwrap() - 8
+                );
+            }
+        }
+    }
+}
+
+pub async fn print_info(attributes: &mut HardwareAttributes) {
+    println!("\n\nx1b[1;32m Kernel Attributes");
```
Comment (george-cosma) [inline tockloader-cli/src/display.rs:215]:
```suggestion
    println!("\n\n\x1b[1;32m Kernel Attributes");
```

File: tockloader-cli/src/display.rs [hunk 8]
```diff
@@ -0,0 +1,242 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright OXIDOS AUTOMOTIVE 2024.
+
+use tockloader_lib::attributes::{
+    app_attributes::AppAttributes, hardware_attributes::HardwareAttributes,
+};
+
+pub async fn print_list(app_details: &mut [AppAttributes], info: bool) {
```
Comment (george-cosma) [inline tockloader-cli/src/display.rs:9]: I would split this into two functions:
`print_list_basic(app_detials)` and
`print_list_detailed(app_details)`

to remove the need for the `info` boolean.

File: tockloader-lib/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    get_apps_data(&mut core)
+}
+
+pub async fn info_probe(
+    choice: DebugProbeInfo,
+    board: &str,
+    chip: &str,
+    core_index: &usize,
+) -> (HardwareAttributes, Vec<AppAttributes>) {
```
Comment (george-cosma) [inline tockloader-lib/src/lib.rs:40]: I would avoid returning pairs. This isn't that bad since it's clear what they are. Please see coment for  the `hardware_attributes.rs` file.

File: tockloader-lib/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+        // Then the bytes that go into the page
+        for b in binary[(i as usize * page_size)..((i + 1) as usize * page_size)].to_vec() {
+            pkt.push(b);
+        }
+
+        // TODO(Micu Ana): Write to bootloader
+    }
+
+    Ok(())
```
Comment (george-cosma) [inline tockloader-lib/src/lib.rs:148]: This is not yet ready. Can be removed for now, until it is finished.

```suggestion
    todo!()
```

File: tockloader-lib/src/bootloader_serial.rs
Comment (george-cosma) [inline tockloader-lib/src/bootloader_serial.rs:1]: I think we can remove this for this PR. Can be re-introduced when we have finished this.

File: tockloader-lib/src/tabs/mod.rs [hunk 11]
```diff
@@ -0,0 +1,6 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright OXIDOS AUTOMOTIVE 2024.
+
+pub mod app_tab;
+pub mod tab;
```
Comment (george-cosma) [inline tockloader-lib/src/tabs/mod.rs:6]: The two mod names aren't that distinct. It's unclear what the difference between them.

File: tockloader-lib/src/tabs/mod.rs
Comment (george-cosma) [inline tockloader-lib/src/tabs/mod.rs:1]: I think these two can be removed as they are unfinished. Eventually, replace function invokations with `todo!()`, or replace the install handler in `main.rs` in the CLI with `todo!()`

**Review by george-cosma — CHANGES_REQUESTED** [review]
Good overall, but a lot of changes! After this PR we need to rework our error handling so the lib has no way to panic, and start to look into logging for the lib. For the CLI, `println`s are good enough, and for error handling we could look into anyhow.

**Comment by george-cosma** [comment]
`/.vscode` has been added back to .gitignore, but the file is still in the PR. It needs to be removed.
https://stackoverflow.com/a/1274447

I'd advise caution when using this command!

File: tock-process-console/src/ui_management/pages/terminal_page/components/terminal_box.rs (hunk 1 above)
Comment (micuami) [reply tock-process-console/src/ui_management/pages/terminal_page/components/terminal_box.rs:17]: The reason I made this modification is because it was suggested when running "cargo build --verbose": 
error[E0308]: mismatched types
   --> tock-process-console/src/ui_management/pages/terminal_page/components/terminal_box.rs:255:82
    |
255 |                 let pseudo_terminal = PseudoTerminal::new(parser.screen()).block(block);
    |                                                                            ----- ^^^^^ expected `ratatui::widgets::block::Block<'_>`, found `ratatui::widgets::Block<'_>`

File: tockloader-lib/Cargo.toml (hunk 2 above)
Comment (micuami) [reply tockloader-lib/Cargo.toml:27]: Removed tar; the other two are used by files from "attributes" folder.

Comment (george-cosma) [reply tockloader-lib/Cargo.toml:27]: Oh, I meant to highlight the entirety of `Cargo.toml`. Could you also check if the other dependencies, besides these 3 higlighted here, are also still needed?

**Review by george-cosma — APPROVED** [review]

**Review by NegrilaRares — APPROVED** [review]
